### PR TITLE
DT-2479: ✨ Retrieve linked account information

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/HomeController.kt [hunk 1]
```diff
@@ -14,15 +14,14 @@ import uk.gov.justice.digital.hmpps.oauth2server.service.UserContextService
...
   fun home(authentication: Authentication): ModelAndView {
-    // special case for azure users - grab all their accounts and combine roles from them too
-    val azureUserDetails = authentication.principal as UserDetailsImpl
-    val authSource = AuthSource.fromNullableString(azureUserDetails.authSource)
-    val authorities = if (authSource == azuread) {
-      val users = userContextService.discoverUsers(authentication.principal as UserDetailsImpl)
-      users.flatMap { it.authorities }
+    val userDetails = authentication.principal as UserDetailsImpl
+    val authSource = AuthSource.fromNullableString(userDetails.authSource)
+    val authorities = if (authSource == AuthSource.azuread || linkAccounts) {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/HomeController.kt:23]: only show extra authorities if `linkAccounts` flag enabled (or azure user)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/account/AccountController.kt [hunk 2]
```diff
@@ -32,17 +33,20 @@ class AccountController(
     @CookieValue(value = "returnTo", defaultValue = "Lw==") returnToFromCookie: String
   ):
     ModelAndView {
-    val username = authentication.name
-    val user = userService.findMasterUserPersonDetails(username).orElseThrow()
-    val userInAuth = userService.getUserWithContacts(username)
+    val userDetails = authentication.principal as UserDetailsImpl
+    val user = userService.getMasterUserPersonDetails(
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/account/AccountController.kt:37]: this is the preferred way of finding users - by using `username` and also auth source

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/UserContextService.kt [hunk 3]
```diff
@@ -10,38 +11,38 @@ import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource.delius
...
   fun discoverUsers(loginUser: UserPersonDetails, scopes: Set<String> = emptySet()): List<UserPersonDetails> {
+    val authSource = AuthSource.fromNullableString(loginUser.authSource)
+    if (authSource != azuread && !linkAccounts) return emptyList()
+
     // if specific accounts are requested via scopes, attempt to find just those.
     // otherwise, attempt to find all accounts.
     val requestedSources = AuthSource.values()
       .filter { it.name in scopes }
-      .let { if (it.isEmpty()) setOf(auth, nomis, delius) else it }
+      .let { it.ifEmpty { setOf(auth, nomis, delius) } }
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/UserContextService.kt:33]: intellij suggestion

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/UserContextService.kt [hunk 4]
```diff
@@ -10,38 +11,38 @@ import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource.delius
...
-    return when (AuthSource.fromNullableString(loginUser.authSource)) {
-      azuread -> {
-        requestedSources
-          .map { mapFromAzureAD(loginUser.userId, it).filter { it.isEnabled } }
-          .filter { it.isNotEmpty() }
-          .flatten()
-      }
-      else -> emptyList()
-    }
+    val email = userService.getEmail(loginUser) ?: return emptyList()
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/UserContextService.kt:35]: the authentication in the session only has the email address for auzre users, for others need to look it up

**Review by SimonMitchellMOJ — APPROVED** [review]
LGTM

**Review by SimonMitchellMOJ — APPROVED** [review]
LGTM

**Pull request merged**